File: hw/fuentes/U3/HW17Maze/NewMazeGenerator/src/main/java/ec/edu/espe/mazegenerator/controller/MazeGenerator.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package ec.edu.espe.mazegenerator.controller;

import ec.edu.espe.mazegenerator.model.EntranceDoor;
import ec.edu.espe.mazegenerator.model.ExitDoor;
import ec.edu.espe.mazegenerator.model.InnerDoor;
import ec.edu.espe.mazegenerator.model.Maze;
import ec.edu.espe.mazegenerator.model.Room;
import ec.edu.espe.mazegenerator.model.Wall;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Random;
import java.util.Set;
import java.util.Stack;

/**
 *
 * @author jorge
 */
public class MazeGenerator {
     public Maze generateMaze(int N, int M) {
        Maze maze = new Maze(N, M);
        addDefaultWalls(maze); 

        Set<Room> visited = new HashSet<>();
        Stack<Room> stack = new Stack<>();

        Room start = maze.getRoom(0, 0);
        Room end = maze.getRoom(N - 1, M - 1);

        maze.setEntrance(start);
        maze.setExit(end);

        visited.add(start);
        stack.push(start);

        while (!stack.isEmpty()) {
            Room current = stack.peek();
            List<Room> neighbors = getUnvisitedNeighbors(maze, current, visited);

            if (!neighbors.isEmpty()) {
                Room next = neighbors.get(new Random().nextInt(neighbors.size()));
                removeWallBetween(current, next);
                createInnerDoor(current, next);   
                visited.add(next);
                stack.push(next);
            } else {
                stack.pop();
            }
        }

        start.setDoor(new EntranceDoor(start));
        end.setDoor(new ExitDoor(end));

        return maze;
    }

    private void addDefaultWalls(Maze maze) {
        for (int i = 0; i < Maze.N; i++) {
            for (int j = 0; j < Maze.M; j++) {
                Room room = maze.getRoom(i, j);
                room.getWalls().clear();
                room.addWall(new Wall('N', true));
                room.addWall(new Wall('S', true));
                room.addWall(new Wall('E', true));
                room.addWall(new Wall('W', true));
            }
        }
    }

    private void removeWallBetween(Room a, Room b) {
        int dx = b.getX() - a.getX();
        int dy = b.getY() - a.getY();

        if (dx == 1) { 
            removeWall(a, 'S');
            removeWall(b, 'N');
        } else if (dx == -1) { 
            removeWall(a, 'N');
            removeWall(b, 'S');
        } else if (dy == 1) { 
            removeWall(a, 'E');
            removeWall(b, 'W');
        } else if (dy == -1) { 
            removeWall(a, 'W');
            removeWall(b, 'E');
        }
    }

    private void removeWall(Room room, char direction) {
        room.getWalls().removeIf(w -> w.getDirection() == direction);
    }

    private List<Room> getUnvisitedNeighbors(Maze maze, Room room, Set<Room> visited) {
        List<Room> neighbors = new ArrayList<>();
        int x = room.getX();
        int y = room.getY();

        if (x > 0 && !visited.contains(maze.getRoom(x - 1, y))) neighbors.add(maze.getRoom(x - 1, y));
        if (x < Maze.N - 1 && !visited.contains(maze.getRoom(x + 1, y))) neighbors.add(maze.getRoom(x + 1, y));
        if (y > 0 && !visited.contains(maze.getRoom(x, y - 1))) neighbors.add(maze.getRoom(x, y - 1));
        if (y < Maze.M - 1 && !visited.contains(maze.getRoom(x, y + 1))) neighbors.add(maze.getRoom(x, y + 1));

        return neighbors;
    }

    private void createInnerDoor(Room from, Room to) {
        from.setDoor(new InnerDoor(from, to));
        to.setDoor(new InnerDoor(to, from));
    }
}

File: hw/fuentes/U3/HW17Maze/MazeAppPY/ec/edu/espe/Maze/controller/MazeGenerator.py
import random
from model.Maze import Maze
from model.Wall import Wall
from model.EntranceDoor import EntranceDoor
from model.ExitDoor import ExitDoor
from model.InnerDoor import InnerDoor

class MazeGenerator:
    def generate_maze(self, n, m):
        maze = Maze(n, m)
        self.add_default_walls(maze)

        visited = set()
        stack = []

        start = maze.get_room(0, 0)
        end = maze.get_room(n - 1, m - 1)

        maze.set_entrance(start)
        maze.set_exit(end)

        visited.add((start.get_x(), start.get_y()))
        stack.append(start)

        while stack:
            current = stack[-1]
            neighbors = self.get_unvisited_neighbors(maze, current, visited)

            if neighbors:
                next_room = random.choice(neighbors)
                self.remove_wall_between(current, next_room)
                self.create_inner_door(current, next_room)
                visited.add((next_room.get_x(), next_room.get_y()))
                stack.append(next_room)
            else:
                stack.pop()

        start.set_door(EntranceDoor(start))
        end.set_door(ExitDoor(end))

        return maze

    def add_default_walls(self, maze):
        for i in range(maze.N):
            for j in range(maze.M):
                room = maze.get_room(i, j)
                room.get_walls().clear()
                room.add_wall(Wall('N', True))
                room.add_wall(Wall('S', True))
                room.add_wall(Wall('E', True))
                room.add_wall(Wall('W', True))

    def remove_wall_between(self, a, b):
        dx = b.get_x() - a.get_x()
        dy = b.get_y() - a.get_y()
        if dx == 1:
            self.remove_wall(a, 'S')
            self.remove_wall(b, 'N')
        elif dx == -1:
            self.remove_wall(a, 'N')
            self.remove_wall(b, 'S')
        elif dy == 1:
            self.remove_wall(a, 'E')
            self.remove_wall(b, 'W')
        elif dy == -1:
            self.remove_wall(a, 'W')
            self.remove_wall(b, 'E')

    def remove_wall(self, room, direction):
        room.walls = [w for w in room.get_walls() if w.get_direction() != direction]

    def get_unvisited_neighbors(self, maze, room, visited):
        neighbors = []
        x, y = room.get_x(), room.get_y()

        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            neighbor = maze.get_room(nx, ny)
            if neighbor and (nx, ny) not in visited:
                neighbors.append(neighbor)

        return neighbors

    def create_inner_door(self, from_room, to_room):
        from_room.set_door(InnerDoor(from_room, to_room))
        to_room.set_door(InnerDoor(to_room, from_room))
